### PR TITLE
Story#510

### PR DESCRIPTION
**Pull request by rakeshSgr**

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -52,4 +52,9 @@ PUNJAB_SERVICE_BASE_URL = "" // Full http path
 PUNJAB_SERVICE_KEY = "" // Access Key
 PUNJAB_SERVICE_HOST = "" // Host URL
 PUNJAB_SERVICE_DEFAULT_PASSWORD = "" // Default Password to create Punjab Users
-PUNJAB_SERVICE_DEFAULT_MAIL_DOMAIN = "" // Default mail domain for Punjab Users
\ No newline at end of file
+PUNJAB_SERVICE_DEFAULT_MAIL_DOMAIN = "" // Default mail domain for Punjab Users
+
+
+SUNBIRD_ORGANISATION_ID = "";
+sunbird_url ="";
```
Comment (aks30) [inline .env.sample]: @rakeshSgr - Please keep all constants in Upper Case

File: controllers/v1/platformUserRolesController.js [hunk 2]
```diff
@@ -190,4 +190,31 @@ module.exports = class PlatformUserRoles extends Abstract {
...
   }
 
+  /**
+  * @api {post} /user-management-service/api/v1/platformUserRoles/create add User
+  * @apiVersion 1.0.0
+  * @apiName create User Roles
+  * @apiGroup Platform User Role Extension
+  * @apiParam {reqeuestBody} consist of body of the request
+  * @apiSampleRequest /user-management-service/api/v1/platformUserRoles/create
+  * @apiUse successBody
```
Comment (aks30) [inline controllers/v1/platformUserRolesController.js:201]: @rakeshSgr - Please do add API request body and response body

File: controllers/v1/platformUserRolesController.js [hunk 3]
```diff
@@ -190,4 +190,31 @@ module.exports = class PlatformUserRoles extends Abstract {
     })
   }
 
+  /**
+  * @api {post} /user-management-service/api/v1/platformUserRoles/create add User
+  * @apiVersion 1.0.0
+  * @apiName create User Roles
```
Comment (aks30) [inline controllers/v1/platformUserRolesController.js]: @rakeshSgr - create should be Create. Also is this API add User or create User Roles

File: generics/message-constants/endpoints.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+ * name : messages-constants/endpoints.js
+ * author : Rakesh Kumar
+ * Date : 18-March-2020
+ * Description : All service endpoints
+ */
+
+module.exports = {
+    SUNBIRD_CREATE_USER  : "/api/user/v1/create",
+    SUNBIRD_ADD_USER_TO_ORG : "/api/org/v1/member/add"
+
```
Comment (aks30) [inline generics/message-constants/endpoints.js]: @rakeshSgr - Remove extra spaces please.

File: generics/services/sunbird.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  * @param token - Logged in user token.
+  * @returns {Promise}
+*/
+
+var addUserToOrganisation = async function ( requestBody,token ) {
+
+    const adduserToOrgUrl = 
+    process.env.sunbird_url+messageConstants.endpoints.SUNBIRD_ADD_USER_TO_ORG;
+
+    return new Promise(async (resolve,reject)=>{
```
Comment (aks30) [inline generics/services/sunbird.js:165]: @rakeshSgr - Can this common code between 2 functions be made the same, please

File: generics/services/sunbird.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+var createUser = async function ( requestBody,token ) {
+
+    const createUserUrl = 
+    process.env.sunbird_url+messageConstants.endpoints.SUNBIRD_CREATE_USER;
+
+    return new Promise(async (resolve,reject)=>{
+        
+        let jsonObject = {
+            firstName:requestBody.firstName,
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - No gender added yet ?

File: models/platformRoles.js [hunk 7]
```diff
@@ -6,6 +6,7 @@ module.exports = {
     createdBy : String,
     updatedBy : String,
     status : String,
-    isDeleted: Boolean
+    isDeleted: Boolean,
```
Comment (aks30) [inline models/platformRoles.js]: @rakeshSgr - Let's add proper validations like default value, also code and title, etc cannot be blank

File: models/platformUserRoles.js [hunk 8]
```diff
@@ -7,6 +7,9 @@ module.exports = {
     userId : String,
     username : String,
     status : String,
-    isDeleted: Boolean
+    isDeleted: Boolean,
+    metaInformation:Object,
+    organisations:Array,
+    organisationRoles:Array
```
Comment (aks30) [inline models/platformUserRoles.js]: @rakeshSgr - Keep consistent spacing between key and value please.

File: module/platformUserRoles/helper.js [hunk 9]
```diff
@@ -185,4 +187,93 @@ module.exports = class platformUserRolesHelper {
 
     }
 
+    
+    static create(request, token) {
```
Comment (aks30) [inline module/platformUserRoles/helper.js]: @rakeshSgr - Please do mention parameter names that make sense and comments on top of a function that aptly describes the same.

File: module/platformUserRoles/helper.js [hunk 10]
```diff
@@ -185,4 +187,93 @@ module.exports = class platformUserRolesHelper {
...
+
+                    
+                    let rolesId = [];
+                    let organisationsRoles= [];
+                    let plaformRoles = [];
+                    let allRoles =[];
+
+                    await Promise.all(request.body.roles.map(async function(roleInfo){
+
+                        let rolesDocument = await database.models.platformRolesExt.findOne({ _id:roleInfo.value },
```
Comment (aks30) [inline module/platformUserRoles/helper.js]: @rakeshSgr - Why query multiple times? Can it not be a single query?

File: module/platformUserRoles/helper.js [hunk 11]
```diff
@@ -185,4 +187,93 @@ module.exports = class platformUserRolesHelper {
...
+                                     rolesDocument.platformRole==true ){
+
+                                   
+                                    plaformRoles.push(rolesDocument.code);
+                                }
+
+                            }
+                    }));
+                  
+                    let object ={
```
Comment (aks30) [inline module/platformUserRoles/helper.js]: @rakeshSgr - Please provide proper object names that make sense.

File: module/platformUserRoles/helper.js [hunk 12]
```diff
@@ -185,4 +187,93 @@ module.exports = class platformUserRolesHelper {
...
+                            "userId": response.result.userId,
+                            "organisationId": request.body.organisations.value,
+                            "roles": plaformRoles
+                        }
+                      
+                        organisationsRoles.push({ organisationId:request.body.organisations.value,roles:rolesId });
+                        let addUserToOrg = await sunBirdService.addUserToOrganisation(object,request.userDetails.userToken);
```
Comment (aks30) [inline module/platformUserRoles/helper.js]: @rakeshSgr - If the token is explicitly passed here in the function why use it again from request. Fix would be to not pass entire request here but only the body and token as 2 arguments

Comment (aks30) [inline module/platformUserRoles/helper.js]: @rakeshSgr - How are we handling the failure from sunbird API?

File: generics/services/sunbird.js (hunk 6 above)
Comment (rakeshSgr) [reply generics/services/sunbird.js]: I have in my local code

File: models/platformRoles.js (hunk 7 above)
Comment (rakeshSgr) [reply models/platformRoles.js]: this is an already created, model. I will update the changes.

Comment (rakeshSgr) [reply models/platformRoles.js]: what is the default values for code and title ?

**Comment by aks30** [comment]
No longer needed

**Pull request closed without merging**